{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a1e47f30",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "4177f988",
   "metadata": {},
   "outputs": [],
   "source": [
    "#OBTAIN THE LIST OF FILE IN THE NUCLEOTIDE FOLDER\n",
    "path = '/Users/sebastiansolano/Documents/AI_projects/HANN_Oncogene/TCGA_project_data_for_training/CPTAC-3_Brain-nos_sinlge_nucleotide_variation/'\n",
    "file_pattern = os.path.join(path, '*')\n",
    "all_files = glob.glob(file_pattern)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "1beac408",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CREATE A LIST WITH AL DATASET\n",
    "df_list = (pd.read_csv(file, sep='\\t') for file in all_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "f156e8eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/d9/jl_fsgyd0hvb0vfnm2kxzjfr0000gn/T/ipykernel_2912/2887450621.py:2: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  combined_df = pd.concat(df_list, ignore_index=True)\n"
     ]
    }
   ],
   "source": [
    "#CONCATENATE THE DATASETS\n",
    "combined_df = pd.concat(df_list, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "d2952d33",
   "metadata": {},
   "outputs": [],
   "source": [
    "#LIST THE CATEGORIES IN VARIAN CALSSIFICATION\n",
    "variant_category=combined_df[\"Variant_Classification\"].unique()\n",
    "variant_category=variant_category.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d37fa11",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ELIMMINATE THE NOT NECCESARY VARIANT CATEGORIES\n",
    "remove_list=[\"3'Flank\",\"5'Flank\",'RNA','Silent']\n",
    "for remove in remove_list:\n",
    "    variant_category.remove(remove)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "00501179",
   "metadata": {},
   "outputs": [],
   "source": [
    "#FILTER THE DATABSE TO WORK ONLY WITH THE \"MAIN\" VARIANT CATEGORIES \n",
    "Variant_dataset_trainig=combined_df[combined_df[\"Variant_Classification\"].isin(variant_category)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe6720ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>...</th>\n",
       "      <th>normal_bam_uuid</th>\n",
       "      <th>case_id</th>\n",
       "      <th>GDC_FILTER</th>\n",
       "      <th>COSMIC</th>\n",
       "      <th>hotspot</th>\n",
       "      <th>RNA_Support</th>\n",
       "      <th>RNA_depth</th>\n",
       "      <th>RNA_ref_count</th>\n",
       "      <th>RNA_alt_count</th>\n",
       "      <th>callers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows Ã— 140 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Hugo_Symbol, Entrez_Gene_Id, Center, NCBI_Build, Chromosome, Start_Position, End_Position, Strand, Variant_Classification, Variant_Type, Reference_Allele, Tumor_Seq_Allele1, Tumor_Seq_Allele2, dbSNP_RS, dbSNP_Val_Status, Tumor_Sample_Barcode, Matched_Norm_Sample_Barcode, Match_Norm_Seq_Allele1, Match_Norm_Seq_Allele2, Tumor_Validation_Allele1, Tumor_Validation_Allele2, Match_Norm_Validation_Allele1, Match_Norm_Validation_Allele2, Verification_Status, Validation_Status, Mutation_Status, Sequencing_Phase, Sequence_Source, Validation_Method, Score, BAM_File, Sequencer, Tumor_Sample_UUID, Matched_Norm_Sample_UUID, HGVSc, HGVSp, HGVSp_Short, Transcript_ID, Exon_Number, t_depth, t_ref_count, t_alt_count, n_depth, n_ref_count, n_alt_count, all_effects, Allele, Gene, Feature, Feature_type, One_Consequence, Consequence, cDNA_position, CDS_position, Protein_position, Amino_acids, Codons, Existing_variation, DISTANCE, TRANSCRIPT_STRAND, SYMBOL, SYMBOL_SOURCE, HGNC_ID, BIOTYPE, CANONICAL, CCDS, ENSP, SWISSPROT, TREMBL, UNIPARC, UNIPROT_ISOFORM, RefSeq, MANE, APPRIS, FLAGS, SIFT, PolyPhen, EXON, INTRON, DOMAINS, 1000G_AF, 1000G_AFR_AF, 1000G_AMR_AF, 1000G_EAS_AF, 1000G_EUR_AF, 1000G_SAS_AF, ESP_AA_AF, ESP_EA_AF, gnomAD_AF, gnomAD_AFR_AF, gnomAD_AMR_AF, gnomAD_ASJ_AF, gnomAD_EAS_AF, gnomAD_FIN_AF, gnomAD_NFE_AF, gnomAD_OTH_AF, gnomAD_SAS_AF, MAX_AF, MAX_AF_POPS, gnomAD_non_cancer_AF, ...]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 140 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#VERIFY NO SILENT MUTATION WERE INCLUDE\n",
    "Variant_dataset_trainig[Variant_dataset_trainig['Variant_Classification']=='Silent']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "34383197",
   "metadata": {},
   "outputs": [],
   "source": [
    "#SAVE THE FILE\n",
    "Variant_dataset_trainig.to_csv('/Users/sebastiansolano/Documents/AI_projects/HANN_Oncogene/TCGA_project_data_for_training/CPTAC-3_Brain-nos_sinlge_nucleotide_variation/unifited_mutation.csv',index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vsc-data-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
