{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "358f57fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import glob\n",
    "from pathlib import Path\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "07ac1b7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Upload the Reactome to ENSEMBL database and get the data for human\n",
    "reactome= pd.read_csv(\"/Users/sebastiansolano/Documents/AI_projects/HANN_Oncogene/Reactome_for_hyperedges/Ensembl2Reactome_All_Levels.txt\", sep='\\t')\n",
    "reactome.columns = ['Gene','Pathway','Pathway_link','Description',\"Evidence\",\"Especie\"]\n",
    "reactome_human =  reactome[reactome[\"Especie\"] == 'Homo sapiens']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c9cc537",
   "metadata": {},
   "source": [
    "# Create the unified transcriptomic Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b065637b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get the list of transcriptomic file\n",
    "path = '/Users/sebastiansolano/Documents/AI_projects/HANN_Oncogene/TCGA_project_data_for_training/CPTAC-3_Brain-nos_transcriptome_profiling'\n",
    "file_pattern = os.path.join(path, '*.tsv')\n",
    "all_files_rnaseq = glob.glob(file_pattern)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5ff0d917",
   "metadata": {},
   "outputs": [],
   "source": [
    "#The following code Merge the files in one single file\n",
    "list_transcriptomic_brain_tumor=[]\n",
    "for filename in all_files_rnaseq:\n",
    "    df = pd.read_csv(filename,sep='\\t')\n",
    "    df.columns=['gene_id','gene_name','gene_type','unstranded','stranded_first','stranded_second','tpm_unstranded','fpkm_unstranded','fpkm_uq_unstranded']\n",
    "    df['ensembl_id']= df['gene_id'].apply(lambda x: x.split(\".\")[0])\n",
    "    df_filter=df\n",
    "    \n",
    "    dataframe_to_list=df_filter[['ensembl_id','unstranded']]\n",
    "    name_in_colum= Path(filename).stem\n",
    "    dataframe_to_list=dataframe_to_list.rename(columns={'unstranded':name_in_colum})\n",
    "    dataframe_to_list= dataframe_to_list.sort_values(by='ensembl_id')\n",
    "    dataframe_to_list= dataframe_to_list.set_index('ensembl_id')\n",
    "    dataframe_to_list= dataframe_to_list.groupby(level=0).sum()\n",
    "    list_transcriptomic_brain_tumor.append(dataframe_to_list)\n",
    "\n",
    "#merge \n",
    "merged_df = pd.concat(list_transcriptomic_brain_tumor, axis=1, join='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f87c0bbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#There are some genes from Reactome databse that were not cover by the transcirptomic analysis\n",
    "not_in_lista =[]\n",
    "for x in reactome_human['Gene']:\n",
    "    if x not in merged_df.index:\n",
    "        not_in_lista.append(x)\n",
    "\n",
    "print(not_in_lista)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9b49023a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We must be sure to analyse the network with the same ammout of genes, that is why we again filter the Reactome database to obtain genes present in both datasets\n",
    "final_reactome_toHYperedges = reactome_human[reactome_human['Gene'].isin(merged_df.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "87a09d96",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save files\n",
    "final_reactome_toHYperedges.to_csv('/Users/sebastiansolano/Documents/AI_projects/HANN_Oncogene/Reactome_for_hyperedges/reactome_patways.csv')\n",
    "merged_df.to_csv('/Users/sebastiansolano/Documents/AI_projects/HANN_Oncogene/TCGA_project_data_for_training/CPTAC-3_Brain-nos_transcriptome_profiling/unified_transcriptomic.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38bd1a04",
   "metadata": {},
   "source": [
    "# Create the Incidence Matrix and Degree matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "65f788bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load transcriptome file\n",
    "merged_df= pd.read_csv('/Users/sebastiansolano/Documents/AI_projects/HANN_Oncogene/TCGA_project_data_for_training/CPTAC-3_Brain-nos_transcriptome_profiling/unified_transcriptomic.csv',index_col='ensembl_id')\n",
    "final_reactome_toHYperedges = pd.read_csv('/Users/sebastiansolano/Documents/AI_projects/HANN_Oncogene/Reactome_for_hyperedges/reactome_patways.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fb585fbd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/d9/jl_fsgyd0hvb0vfnm2kxzjfr0000gn/T/ipykernel_2795/1707008027.py:3: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  incidence_matrix = incidence_matrix.fillna(0)\n"
     ]
    }
   ],
   "source": [
    "#Incidence matrix\n",
    "incidence_matrix = pd.DataFrame(index=merged_df.index, columns=list(set(final_reactome_toHYperedges['Pathway'])))\n",
    "incidence_matrix = incidence_matrix.fillna(0)\n",
    "\n",
    "#Fill up the matrix\n",
    "for z in merged_df.index:\n",
    "    sub_set= final_reactome_toHYperedges[final_reactome_toHYperedges['Gene']==z]\n",
    "    list_pathways= sub_set['Pathway'].to_list()\n",
    "    for c in list_pathways:\n",
    "        incidence_matrix.loc[z,c]=1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bb15ac2",
   "metadata": {},
   "source": [
    "# Build the Variant Vector for Each Gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0bfa62b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the transcriptomic file and the variant file\n",
    "Genes= pd.read_csv('/Users/sebastiansolano/Documents/AI_projects/HANN_Oncogene/TCGA_project_data_for_training/CPTAC-3_Brain-nos_transcriptome_profiling/unified_transcriptomic.csv')\n",
    "variant = pd.read_csv('/Users/sebastiansolano/Documents/AI_projects/HANN_Oncogene/TCGA_project_data_for_training/CPTAC-3_Brain-nos_sinlge_nucleotide_variation/unified_mutation.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "074b7636",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create dataframe filled up with zeros\n",
    "Variant_dataframe_values= pd.DataFrame(index=Genes['ensembl_id'], columns=list(set(variant['Variant_Classification'])))\n",
    "Variant_dataframe_values=Variant_dataframe_values.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "13bfc2fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#fill up the dataframe with the freqeuncy of each kind of variant clasiffication per each gene\n",
    "\n",
    "for gene in Variant_dataframe_values.index:\n",
    "    sub_set= variant[variant['Gene'] == gene]\n",
    "    for var in Variant_dataframe_values.columns:\n",
    "        sub_path= sub_set[sub_set['Variant_Classification']==var]\n",
    "        frequency= sub_path.shape[0]\n",
    "        Variant_dataframe_values.loc[gene,var]=frequency\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed8e28be",
   "metadata": {},
   "source": [
    "# Normalize the transcriptomic Dataset and Merge with the Variant dataset to set the final vector matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "74734f00",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7222c18e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create Metadata \n",
    "metadata = pd.DataFrame({\n",
    "    'condition': ['control'] * merged_df.shape[1]\n",
    "}, index=merged_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0be44548",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/d9/jl_fsgyd0hvb0vfnm2kxzjfr0000gn/T/ipykernel_3850/1212740904.py:2: DeprecationWarning: design_factors is deprecated and will soon be removed.Please consider providing a formulaic formula using the design argumentinstead.\n",
      "  dds = DeseqDataSet(\n"
     ]
    }
   ],
   "source": [
    "# Initialize DeseqDataSet\n",
    "dds = DeseqDataSet(\n",
    "    counts=merged_df.T,\n",
    "    metadata=metadata,\n",
    "    design_factors='condition',  # your experimental design\n",
    "    refit_cooks=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55599980",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Normalize samples\n",
    "dds.fit_size_factors()\n",
    "dds.fit_genewise_dispersions()\n",
    "dds.vst()\n",
    "\n",
    "vst_counts = dds.layers['vst_counts']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7fb15ccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert back to DataFrame (still samples x genes)\n",
    "normalized_df = pd.DataFrame(\n",
    "    vst_counts,\n",
    "    index=merged_df.columns,  # sample names\n",
    "    columns= merged_df.index  # gene names\n",
    ").T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80e41f08",
   "metadata": {},
   "source": [
    "# Merge Variant Calling and Transcriptome Profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a92ac8e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Merge Variant Calling and Transcriptome profile\n",
    "Feature_dataframe = Variant_dataframe_values.join(normalized_df,how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "35c8bbbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make both incidence and feature matrices have the same row order\n",
    "\n",
    "Index_order=Feature_dataframe.index.to_list()\n",
    "incidence_matrix=incidence_matrix.loc[Index_order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07bcc26a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save incidence Matrix and Feature Matrix\n",
    "Feature_dataframe.to_csv(\"/Users/sebastiansolano/Documents/AI_projects/HANN_Oncogene/Input_HGNN/Feature_matrix.csv\")\n",
    "incidence_matrix.to_csv(\"/Users/sebastiansolano/Documents/AI_projects/HANN_Oncogene/Input_HGNN/Incidence_matrix.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vsc-data-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
